voicevox_engine/sv_model.py

```diff
@@ -0,0 +1,115 @@
+import base64
+import json
+import os
+import shutil
+from pathlib import Path
+from typing import List
+
+from appdirs import user_data_dir
+
+from voicevox_engine.model import SVModelInfo
+
+# FIXME: ファイル保存場所をエンジン固有のIDが入ったものにする
+save_dir = Path(user_data_dir("sharevox-engine"))
+if not save_dir.is_dir():
+    save_dir.mkdir(parents=True)
+model_dir = save_dir.joinpath("model")
+
+
+def get_all_sv_models(libraries_dir: Path = model_dir) -> List[str]:
+    """
+    保存されているsv_modelsの情報をListで返却する。
+
+    - libraries.jsonを読み込み、key(str)のみをListに詰めて返却する
+    - bool値は使わないので返却しない
+    - permission deniedされたら
+    """
+
+    libraries = None
+    try:
+        with open(libraries_dir.joinpath("libraries.json"), "r") as f:
+            libraries = json.load(f)
+    except Exception as e:
+        raise e
+    return list(libraries.keys())
+
+
+def register_sv_model(sv_model: SVModelInfo, stored_dir: Path = save_dir):
+    """
+    送られた単一のSVModelを保存する。返り値はない。
+    """
+
+    try:
+        # 既存のsv_modelsとUUIDの重複がないかを、ディレクトリ作成時に確認する
+        # 以下のディレクトリを作成する
+        # - /model/${uuid}
+        # - /speaker_info/${uuid}
+        # - /speaker_info/${uuid}/icons/
+        # - /speaker_info/${uuid}/voice_samples/
+        # FileExistsExceptionが起きたら、ここの階層でもraiseして呼び出し元に流す
+        # この後の処理で何らかのExceptionが起きた場合、上記のディレクトリを全て削除する
+
+        # 意味的にはmodelsの方が正しそうだけど、実際に保存されたディレクトリ名はmodelだったので
+        model_uuid_dir = stored_dir.joinpath("model", sv_model.uuid)
+        os.makedirs(model_uuid_dir)
+
+        # variance_model, embedder_model, decoder_modelは
+        # それぞれbase64デコードしてから/model/${uuid}/*.onnxに保存する
+        with open(model_uuid_dir.joinpath("variance_model.onnx"), "wb") as f:
+            f.write(base64.b64decode(sv_model.variance_model.encode("utf-8")))
+        with open(model_uuid_dir.joinpath("embedder_model.onnx"), "wb") as f:
+            f.write(base64.b64decode(sv_model.embedder_model.encode("utf-8")))
+        with open(model_uuid_dir.joinpath("decoder_model.onnx"), "wb") as f:
+            f.write(base64.b64decode(sv_model.decoder_model.encode("utf-8")))
```

Comment:
```suggestion
        with open(model_uuid_dir / "variance_model.onnx", "wb") as f:
            f.write(base64.b64decode(sv_model.variance_model.encode("utf-8")))
        with open(model_uuid_dir / "embedder_model.onnx", "wb") as f:
            f.write(base64.b64decode(sv_model.embedder_model.encode("utf-8")))
        with open(model_uuid_dir / "decoder_model.onnx", "wb") as f:
            f.write(base64.b64decode(sv_model.decoder_model.encode("utf-8")))
```
直感的かつ見通しが良いのでこうしちゃいましょう！(他のjoinpathのところもお願いしたいです！)

Reply:
fix with https://github.com/SHAREVOX/sharevox_engine/pull/1/commits/3d5edb8b70e4c9cbb19d5eba6f1b7f5d92fc8c91